File: learning/deck_order.py
"""Canonical tarot deck ordering.

The crawled JSON is stored in alphabetical order because that is how the
source site lists the cards. Anywhere a human browses the deck we want the
traditional order instead: Major Arcana 0-21, then each suit from Ace to
King. Card names are English in both the EN and VI datasets, so a single
lookup table covers both languages.
"""

from __future__ import annotations

MAJOR_ARCANA_ORDER = [
    "The Fool",
    "The Magician",
    "The High Priestess",
    "The Empress",
    "The Emperor",
    "The Hierophant",
    "The Lovers",
    "The Chariot",
    "Strength",
    "The Hermit",
    "The Wheel of Fortune",
    "Justice",
    "The Hanged Man",
    "Death",
    "Temperance",
    "The Devil",
    "The Tower",
    "The Star",
    "The Moon",
    "The Sun",
    "Judgement",
    "The World",
]

SUIT_ORDER = ["Wands", "Cups", "Swords", "Pentacles"]

RANK_ORDER = [
    "Ace",
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
    "Page",
    "Knight",
    "Queen",
    "King",
]

_MAJOR_INDEX = {
    name.casefold(): index for index, name in enumerate(MAJOR_ARCANA_ORDER)
}
_SUIT_INDEX = {name.casefold(): index for index, name in enumerate(SUIT_ORDER)}
_RANK_INDEX = {name.casefold(): index for index, name in enumerate(RANK_ORDER)}

# Major Arcana first, then the four suits.
_MAJOR_GROUP = 0
_MINOR_GROUP_OFFSET = 1
# Unknown names sort last instead of raising, so a data typo degrades to an
# alphabetical tail rather than breaking the menu.
_UNKNOWN_GROUP = _MINOR_GROUP_OFFSET + len(SUIT_ORDER)


def deck_position(name: str) -> tuple[int, int, str]:
    """Sort key placing a card name in traditional deck order."""
    cleaned = " ".join((name or "").split())
    folded = cleaned.casefold()

    major = _MAJOR_INDEX.get(folded)
    if major is not None:
        return (_MAJOR_GROUP, major, folded)

    rank, separator, suit = cleaned.partition(" of ")
    if separator:
        suit_index = _SUIT_INDEX.get(suit.casefold())
        rank_index = _RANK_INDEX.get(rank.casefold())
        if suit_index is not None and rank_index is not None:
            return (_MINOR_GROUP_OFFSET + suit_index, rank_index, folded)

    return (_UNKNOWN_GROUP, 0, folded)


def sort_cards(cards: list[dict]) -> list[dict]:
    """Return cards ordered by traditional deck position."""
    return sorted(cards, key=lambda card: deck_position(card.get("name") or ""))

File: mobile/lib/src/models/deck_order.dart
/// Traditional tarot deck ordering, mirroring `learning/deck_order.py`.
///
/// The bundled JSON is alphabetical because the source site lists it that way.
/// Card names are English in both the EN and VI datasets, so one table covers
/// both languages.
const List<String> majorArcanaOrder = <String>[
  'The Fool',
  'The Magician',
  'The High Priestess',
  'The Empress',
  'The Emperor',
  'The Hierophant',
  'The Lovers',
  'The Chariot',
  'Strength',
  'The Hermit',
  'The Wheel of Fortune',
  'Justice',
  'The Hanged Man',
  'Death',
  'Temperance',
  'The Devil',
  'The Tower',
  'The Star',
  'The Moon',
  'The Sun',
  'Judgement',
  'The World',
];

const List<String> suitOrder = <String>['Wands', 'Cups', 'Swords', 'Pentacles'];

const List<String> rankOrder = <String>[
  'Ace',
  'Two',
  'Three',
  'Four',
  'Five',
  'Six',
  'Seven',
  'Eight',
  'Nine',
  'Ten',
  'Page',
  'Knight',
  'Queen',
  'King',
];

/// Sort key for a card name: group first (Major Arcana, then each suit),
/// then position inside that group.
class DeckPosition implements Comparable<DeckPosition> {
  const DeckPosition(this.group, this.position, this.name);

  final int group;
  final int position;
  final String name;

  @override
  int compareTo(DeckPosition other) {
    if (group != other.group) {
      return group.compareTo(other.group);
    }
    if (position != other.position) {
      return position.compareTo(other.position);
    }
    return name.compareTo(other.name);
  }
}

final Map<String, int> _majorIndex = <String, int>{
  for (int i = 0; i < majorArcanaOrder.length; i++)
    majorArcanaOrder[i].toLowerCase(): i,
};
final Map<String, int> _suitIndex = <String, int>{
  for (int i = 0; i < suitOrder.length; i++) suitOrder[i].toLowerCase(): i,
};
final Map<String, int> _rankIndex = <String, int>{
  for (int i = 0; i < rankOrder.length; i++) rankOrder[i].toLowerCase(): i,
};

const int _majorGroup = 0;
const int _minorGroupOffset = 1;
// Unknown names sort last instead of throwing, so a data typo degrades to an
// alphabetical tail rather than breaking the browser.
final int _unknownGroup = _minorGroupOffset + suitOrder.length;

DeckPosition deckPositionOf(String name) {
  final String cleaned = name.replaceAll(RegExp(r'\s+'), ' ').trim();
  final String folded = cleaned.toLowerCase();

  final int? major = _majorIndex[folded];
  if (major != null) {
    return DeckPosition(_majorGroup, major, folded);
  }

  final int separator = cleaned.indexOf(' of ');
  if (separator > 0) {
    final int? rank =
        _rankIndex[cleaned.substring(0, separator).toLowerCase()];
    final int? suit =
        _suitIndex[cleaned.substring(separator + 4).toLowerCase()];
    if (rank != null && suit != null) {
      return DeckPosition(_minorGroupOffset + suit, rank, folded);
    }
  }

  return DeckPosition(_unknownGroup, 0, folded);
}
